### DISCUSSION TITLE
Lens grating lens spectrometer

### DISCUSSION THREAD

**Discussion by chalongratken** [mjhoptics/ray-optics#156] [Q&A]
Recently, I am trying to use the rayoptics to model a lens grating lens spectrometer which is basically consisted of a doublet collimator lens, a reflective grating, and a doublet focuser lenses. I  followed the suggestion in #147 to create a ray trace with a collimator lens and a grating which is ok. However, when I tried to add the focuser lens after the grating, the rays seems to be discontinuous and they are diverged instead of focusing as seen in the attached image. Here is my code:

from rayoptics.environment import *  

opm = OpticalModel()   
sm  = opm['seq_model']   
osp = opm['optical_spec']   
pm = opm['parax_model']   
em = opm['ele_model']   
pt = opm['part_tree']   
ar = opm['analysis_results']   
 
osp['pupil'] = PupilSpec(osp, key=['object', 'epd'], value=25.4)  
osp['fov'] = FieldSpec(osp, key=['object', 'height'], flds=[0.0], is_relative=False)
osp['wvls'] = WvlSpec([(400, 0.5), (600, 1.0), (800, 0.5)], ref_wl=1)
opm.radius_mode = True 

sm.gaps[0].thi = 39
# Add Collimator Lens    
sm.add_surface([34.9, 9.0, 'N-BAF10','Schott'], aperture_type='circular', aperture_value=25.0)  
sm.add_surface([-28.8, 3.5, 'N-SF6','Schott'])   
sm.add_surface([-137.5, 0, 'air']) 
sm.gaps[-1].thi = 70.0

# Add grating 
sm.add_surface([.0, -70., 'refl'])  
grating_surface = sm.cur_surface    
sm.ifcs[grating_surface].decenter = srf.DecenterData('bend', alpha=40) 

grating600lp = DiffractionGrating('600L/mm grating', grating_lpmm=600, order=1)
sm.ifcs[grating_surface].phase_element = grating600lp

grating_tilt = 40
total_angle = 60

sm.cur_surface = grating_surface
sm.ifcs[sm.cur_surface].decenter = srf.DecenterData('dec and return', alpha=grating_tilt)
sm.gaps[sm.cur_surface].thi = .0

sm.add_surface([.0, -70.])
sm.ifcs[sm.cur_surface].interact_mode = 'dummy'
sm.ifcs[sm.cur_surface].decenter = srf.DecenterData('decenter', alpha=total_angle)

sm.gaps[-1].thi = -70

# Add Focusor Lens 
sm.set_stop()
sm.add_surface([34.9, 9.0, 'N-BAF10','Schott'], aperture_type='circular', aperture_value=25.0)  
sm.add_surface([-28.8, 3.5, 'N-SF6','Schott'])   
sm.add_surface([-137.5, 0, 'air'])   

sm.gaps[-1].thi = -50

opm.update_model()   

from rayoptics.elem.layout import RayFanBundle
def create_ray_fan_vs_wvl(fig, opt_model, num_rays=21):
    ray_fan_bundles = []
    system_length, start_offset = fig.sl_so
    fov = opt_model['optical_spec']['fov']
    fi = 0
    fld = fov.fields[fi]
    fld_label = fov.index_labels[fi]
    wvls = opt_model['optical_spec']['wvls']
    for wl, clr in zip(wvls.wavelengths, wvls.render_colors):
        rayfan = RayFan(opt_model, f=fld, wl=wl, xyfan='y', num_rays=num_rays,
                        label=fld_label, color=clr)
        rb = RayFanBundle(opt_model, rayfan, start_offset)
        ray_fan_bundles.append(rb)
    return ray_fan_bundles

entity_factory = create_ray_fan_vs_wvl, (opm,), {'num_rays': 3}
eflist = [entity_factory]

layout_plt3 = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, do_draw_rays=False, 
 do_draw_beams=False, do_draw_edge_rays=False, entity_factory_list=eflist).plot()

I am still not very clear how to add the optical element after a grating after which the the system coordinate has rotated. Please help me on this issue. Thanks 

![Figure 2024-09-09 161312](https://github.com/user-attachments/assets/2fe9e195-f2a4-4bb7-8af1-a84870ca1844)

**Comment by mjhoptics** [comment]
Hi @chalongratken,
Thank you for the detailed script documenting this issue. You are correct, adding optics following a mirror can be confusing. 

Ray-optics follows the typical convention imaging optical software uses which is that light propagates in the negative Z direction after an odd number of reflections.

In your model, you entered the doublet following the grating as if light was propagating in the +Z direction. If you look closely at the lens layout you can see the rays go through surface 6 to 8 "backwards", as seen in the image
```
              r            t        medium     mode   zdr      sd
  Obj:     0.000000      39.0000       air             1      0.0000
    1:    34.900000      9.00000   N-BAF10             1      4.8637
    2:   -28.800000      3.50000     N-SF6             1      4.9992
    3:  -137.500000      70.0000       air             1      5.1080
    4:     0.000000      0.00000       air    reflect -1      8.7644
 Stop:     0.000000     -70.0000       air            -1      8.3725
    6:    34.900000      9.00000   N-BAF10            -1      13.531
    7:   -28.800000      3.50000     N-SF6            -1      12.841
    8:  -137.500000     -50.0000       air            -1      11.925
  Img:    -0.000000      0.00000                      -1      30.149
```
![Screenshot 2024-09-11 000009](https://github.com/user-attachments/assets/6b62fa08-d2f7-482e-9c30-d59715815f95)

You can fix this by scaling the relevant interfaces and gaps by -1. Since you entered the airspace thicknesses 5 and 8 as negative values, these don't need to be scaled.
```
sm.ifcs[6].apply_scale_factor(-1)
sm.gaps[6].apply_scale_factor(-1)
sm.ifcs[7].apply_scale_factor(-1)
sm.gaps[7].apply_scale_factor(-1)
sm.ifcs[8].apply_scale_factor(-1)
opm.update_model()
sm.list_model()
```
```
              r            t        medium     mode   zdr      sd
  Obj:     0.000000      39.0000       air             1      0.0000
    1:    34.900000      9.00000   N-BAF10             1      4.8637
    2:   -28.800000      3.50000     N-SF6             1      4.9992
    3:  -137.500000      70.0000       air             1      5.1080
    4:     0.000000      0.00000       air    reflect -1      8.7644
 Stop:     0.000000     -70.0000       air            -1      8.3725
    6:   -34.900000     -9.00000   N-BAF10            -1      13.981
    7:    28.800000     -3.50000     N-SF6            -1      13.648
    8:   137.500000     -50.0000       air            -1      13.185
  Img:     0.000000      0.00000                      -1      3.3211
```
![Figure 2(1)](https://github.com/user-attachments/assets/5a6f5ae5-c99f-4b0f-912a-24d25418ada7)

I wrote a function that takes the surface range and does the scaling.
```
def scale_for_neg_z_dir(sm, idx1, idx2):
    for i in range(idx1, idx2+1):
        try:
            sm.ifcs[i].apply_scale_factor(-1)
            if i < idx2:
                sm.gaps[i].apply_scale_factor(-1)
        except IndexError as ie:
            break
```
For this case, it would be invoked as:
`scale_for_neg_z_dir(sm, 6, 8)`

Hope this helps.

Mike Hayford

**Reply by chalongratken** [reply]
Hi Michael,
Thanks a lot for detailed explanation which help me a lot.

**Comment by chalongratken** [comment]
Now I want to adapt the above lens-grating-lens spectrometer into a lens-grism-lens conflagration, in which a couple of wedge prisms were added before and after the grating as shown in the attached image. This configuration will give an in-line spectrometer. But I still have no idea about adding prism. Could you please help me on how to add a wedge prism?  Thanks in advance.
![4-Figure1-1](https://github.com/user-attachments/assets/2a63ad81-923d-4715-96c6-7520d749ca04)

**Reply by mjhoptics** [reply]
Hello @chalongratken,

This should be possible. Here's an example with made-up data. Use the recently released v0.9.2 ray-optics to get the rendering to look right. 
Hope this helps.
Mike Hayford

```python
%matplotlib inline
```


```python
from rayoptics.environment import *
```


```python
from rayoptics.elem.layout import RayFanBundle
def create_ray_fan_vs_wvl(fig, opt_model, num_rays=21):
    ray_fan_bundles = []
    system_length, start_offset = fig.sl_so
    fov = opt_model['optical_spec']['fov']
    fi = 0
    fld = fov.fields[fi]
    fld_label = fov.index_labels[fi]
    wvls = opt_model['optical_spec']['wvls']
    for wl, clr in zip(wvls.wavelengths, wvls.render_colors):
        rayfan = RayFan(opt_model, f=fld, wl=wl, xyfan='y', num_rays=num_rays,
        label=fld_label, color=clr)
        rb = RayFanBundle(opt_model, rayfan, start_offset)
        ray_fan_bundles.append(rb)
    return ray_fan_bundles
```


```python
opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']

osp['pupil'] = PupilSpec(osp, key=['object', 'epd'], value=25.4)
osp['fov'] = FieldSpec(osp, key=['object', 'height'], flds=[0.0], is_relative=False)
osp['wvls'] = WvlSpec([(400, 0.5), (600, 1.0), (800, 0.5)], ref_wl=1)
opm.radius_mode = True

sm.do_apertures = False
sm.gaps[0].thi = 39

# Add Collimator Lens
sm.add_surface([34.9, 9.0, 'N-BAF10','Schott'], sd=15)
sm.add_surface([-28.8, 3.5, 'N-SF6','Schott'])
sm.add_surface([-137.5, 15, 'air'])
sm.set_stop()
```


### Enter the first prism. 

It is best to use the "dec and return" or "decenter and return" decenter type for this. In this case, only the first surface needs to be tilted.


```python
# Pick a prism angle
prism_angle = 15

# prism #1
sm.add_surface([0, 8, 'N-BAF10','Schott'], sd=15)
sm.ifcs[sm.cur_surface].decenter = srf.DecenterData('dec and return', alpha=prism_angle)

sm.add_surface([0, 3.5, 'N-BAF10','Schott'])
sm.add_surface([0, 4, 'air'])
```

### Add the grating to the exiting surface of the substrate. 

The ray trace relies on the interact_mode of the surface to determine whether to transmit or reflect the diffrated rays.


```python
grating_surface = sm.cur_surface
grating600lp = DiffractionGrating('600L/mm grating', grating_lpmm=600, order=1)
sm.ifcs[grating_surface].phase_element = grating600lp
```

### Enter prism #2 and the focusor lens. 

Again, a decenter and return surface, tilted the opposite direction, is used on the exiting surface of the second prism assembly.


```python
sm.add_surface([0, 3.5, 'N-BAF10','Schott'], sd=15)
sm.add_surface([0, 8, 'N-BAF10','Schott'])
sm.add_surface([0, 15, 'air'])
sm.ifcs[sm.cur_surface].decenter = srf.DecenterData('dec and return', alpha=-prism_angle)

# Add Focusor Lens
sm.add_surface([34.9, 9.0, 'N-BAF10','Schott'], sd=15)
sm.add_surface([-28.8, 3.5, 'N-SF6','Schott'])
sm.add_surface([-137.5, 0, 'air'])

sm.gaps[-1].thi = 50

opm.update_model()

sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000      39.0000       air             1      1.0000
        1:    34.900000      9.00000   N-BAF10             1      15.000
        2:   -28.800000      3.50000     N-SF6             1      1.0000
     Stop:  -137.500000      15.0000       air             1      1.0000
        4:     0.000000      8.00000   N-BAF10             1      15.000
        5:     0.000000      3.50000   N-BAF10             1      1.0000
        6:     0.000000      4.00000       air             1      1.0000
        7:     0.000000      3.50000   N-BAF10             1      15.000
        8:     0.000000      8.00000   N-BAF10             1      1.0000
        9:     0.000000      15.0000       air             1      1.0000
       10:    34.900000      9.00000   N-BAF10             1      15.000
       11:   -28.800000      3.50000     N-SF6             1      1.0000
       12:  -137.500000      50.0000       air             1      1.0000
      Img:     0.000000      0.00000                       1      1.0000



```python
entity_factory = create_ray_fan_vs_wvl, (opm,), {'num_rays': 3}
eflist = [entity_factory]
```


```python
layout_plt1 = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, do_draw_rays=False,
do_draw_beams=False, do_draw_edge_rays=False, entity_factory_list=eflist).plot()
```


     
![output_13_0](https://github.com/user-attachments/assets/cec5c9d5-721d-4cae-aa66-44b254dba560)



### The list_sg method includes the decenter data in the surface listing.


```python
sm.list_sg()
```

                   r               mode              type          y       alpha
                           t           medium
      Obj:      0.00000                 
                         39.0000          air
        1:      34.9000                 
                         9.00000      N-BAF10
        2:     -28.8000                 
                         3.50000        N-SF6
     Stop:     -137.500                 
                         15.0000          air
        4:      0.00000                      dec and return     0.0000     15.000
                         8.00000      N-BAF10
        5:      0.00000                 
                         3.50000      N-BAF10
        6:      0.00000                 
                         4.00000          air
        7:      0.00000                 
                         3.50000      N-BAF10
        8:      0.00000                 
                         8.00000      N-BAF10
        9:      0.00000                      dec and return     0.0000    -15.000
                         15.0000          air
       10:      34.9000                 
                         9.00000      N-BAF10
       11:     -28.8000                 
                         3.50000        N-SF6
       12:     -137.500                 
                         50.0000          air
      Img:      0.00000                 



```python

```

**Comment by chalongratken** [comment]
Thanks Michael! this is really wonderful.
 
However, when I incorporated the code to visualize the spot diagram of the optical system, it gave me the following error: ValueError: zero-size array to reduction operation maximum which has no identity.

My added code is:
spot_plt = plt.figure(dpi=300, FigureClass=SpotDiagramFigure, opt_model=opm, scale_type=Fit.All_Same).plot()

It seems that spot diagram fail to plot when using a grating.

**Reply by mjhoptics** [reply]
This is a problem with the clear apertures set on the surfaces. If you look at the sm.list_model() output, there are a number of surfaces that have a sd (semi-diameter) of 1. This is clipping all the rays when doing a spot diagram.

You can quickly set them to the same value this way:

```
for ifc in sm.ifcs[1:-1]:
    ifc.set_max_aperture(15.)
```
When this is done, the spot diagram works correctly. It was failing because no rays got through and that made it unhappy :-(

Hope this helps.

Mike Hayford

**Reply by chalongratken** [reply]
Hi Michael,

I tried what you suggested and got the spot diagram worked, thanks. 
Now I improved the code a bit further to make the optical system more realistic based on off-the-shelf optics. 
The following is my current code, and the questions will be asked below.

`from rayoptics.environment import *`

Import RayFanBudle for multiwavelength plot
`from rayoptics.elem.layout import RayFanBundle
def create_ray_fan_vs_wvl(fig, opt_model, num_rays=21):
    ray_fan_bundles = []
    system_length, start_offset = fig.sl_so
    fov = opt_model['optical_spec']['fov']
    fi = 0
    fld = fov.fields[fi]
    fld_label = fov.index_labels[fi]
    wvls = opt_model['optical_spec']['wvls']
    for wl, clr in zip(wvls.wavelengths, wvls.render_colors):
        rayfan = RayFan(opt_model, f=fld, wl=wl, xyfan='y', num_rays=num_rays,
        label=fld_label, color=clr)
        rb = RayFanBundle(opt_model, rayfan, start_offset)
        ray_fan_bundles.append(rb)
    return ray_fan_bundles`

Create optical model
`opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']
osp['pupil'] = PupilSpec(osp, key=['object', 'epd'], value=16)
osp['fov'] = FieldSpec(osp, key=['object', 'height'], flds=[0.0], is_relative=False)
osp['wvls'] = WvlSpec([(400, 0.5), (600, 1.0), (800, 0.5)], ref_wl=1)
opm.radius_mode = True`

Add optical component 
Add object distance
`sm.do_apertures = False
sm.gaps[0].thi = 39`

Add Collimator Lens
`sm.add_surface([34.9, 9.0, 'N-BAF10','Schott'], sd=12.7)
sm.add_surface([-28.8, 3.5, 'N-SF6','Schott'], sd=12.7)
sm.add_surface([-137.5, 15, 'air'], sd=12.7)
sm.set_stop()`

Add Prism 1
`prism_angle = 15
sm.add_surface([0, 8, 'N-BAF10','Schott'], sd=12.7)
sm.ifcs[sm.cur_surface].decenter = srf.DecenterData('dec and return', alpha=prism_angle)
sm.add_surface([0, 0, 'N-BAF10','Schott'], sd=12.7)
sm.add_surface([0, 2, 'air'], sd=12.7)`

Add transmission plate
`sm.add_surface([0, 2, 'SF10','Schott'], sd=12.7)
sm.add_surface([0, 2, 'air'], sd=12.7)`

Add Grating to the existing surface
`grating_surface = sm.cur_surface
grating600lp = DiffractionGrating('600L/mm grating', grating_lpmm=600, order=1)
sm.ifcs[grating_surface].phase_element = grating600lp`

Add Prism 2
`sm.add_surface([0, 0, 'N-BAF10','Schott'], sd=12.7)
sm.add_surface([0, 8, 'N-BAF10','Schott'], sd=12.7)
sm.add_surface([0, 15, 'air'], sd=12.7)
sm.ifcs[sm.cur_surface].decenter = srf.DecenterData('dec and return', alpha=-prism_angle)`

Add Focuser Lens
`sm.add_surface([34.9, 9.0, 'N-BAF10','Schott'], sd=12.7)
sm.add_surface([-28.8, 3.5, 'N-SF6','Schott'], sd=12.7)
sm.add_surface([-137.5, 0, 'air'], sd=12.7)`

Add detector
`sm.add_surface([0, 0.0, 'air','Schott'], sd=10) # Add a flat surface for the image plane
opm.update_model()
sm.list_model()`

Plot Layout and Rays
`entity_factory = create_ray_fan_vs_wvl, (opm,), {'num_rays': 3}
eflist = [entity_factory]
layout_plt1 = plt.figure(dpi=300, FigureClass=InteractiveLayout, opt_model=opm, do_draw_rays=False,
do_draw_beams=False, do_draw_edge_rays=False, entity_factory_list=eflist).plot()`

Plot Spot Diagram 
`spot_plt = plt.figure(dpi=300, FigureClass=SpotDiagramFigure, opt_model=opm,
                      scale_type=Fit.All_Same).plot()`

And here is the result:
![Figure 2024-09-29 233850](https://github.com/user-attachments/assets/0c34d4f7-aa3f-4f6c-86c9-9addb90b7c9e)
![Figure 2024-09-29 233914](https://github.com/user-attachments/assets/8f2ca880-8ae1-4363-974f-66524dcf43b0)

Now, the question are:
1. Actually, I'd like to set the material of the transmission plate to 'fused silica' rather than 'SF10' but could not find any fused silica option.
2. Can we plot the spot diagram separately by wavelength? So more detail could be seen in the individual plots. 
3. How can I access the data of the spot diagram plot? I think it could be some sorts of coordinate data of the points that form the spot diagram. So I can use this data for further processing, e.g., calculate for the RMS radius of the spots, or use it as a criteria in an optimization process that smallest spot is preferred.

**Reply by mjhoptics** [reply]
Hello @chalongratken,
I've updated my `opticalglass` package to import materials from the [RefractiveIndex.Info](https://refractiveindex.info/) database. The `opticalglass` package is a dependency of `ray-optics`; you can update to the current version (v1.0.9) using either pip or conda. The [documentation](https://opticalglass.readthedocs.io/en/latest/OG_RefractiveIndexInfo/OG_RefractiveIndexInfo.html#sio2) gives several examples of how to access materials such as SiO2, CaF2, etc. The material object can be used in the add_surface() call instead of SF10.
I'm heading out of town for the week, I'll post some examples addressing your spot diagram questions when I return.
Regards
Mike Hayford

**Reply by chalongratken** [reply]
Thanks Michael! 
Here is the code I made for the particular surface that requires SiO2 material:

`from opticalglass.glassfactory import create_glass
from opticalglass.rindexinfo import summary_plots
SiO2_url = 'https://refractiveindex.info/database/data-nk/main/SiO2/Malitson.yml'
SiO2 = create_glass(SiO2_url, "rindexinfo")
summary_plots(SiO2)
sm.add_surface([0, 2, SiO2], sd=12.7)`

![image](https://github.com/user-attachments/assets/72404520-f30d-4eb3-96c6-d43871434459)
![image](https://github.com/user-attachments/assets/b512b245-83d8-48d7-b15d-9e501774f06c)

Now I think SiO2 material works for me.

Regards,
Chalongrat
